rubicon/objc/types.py

```diff
@@ -18,17 +18,194 @@
 __arm64__ = (_any_arm and __LP64__)
 __arm__ = (_any_arm and not __LP64__)
 
-PyObjectEncoding = b'{PyObject=@}'
 
-
-def encoding_for_ctype(vartype):
-    typecodes = {
-        c_char: b'c', c_int: b'i', c_short: b's', c_long: b'l', c_longlong: b'q',
-        c_ubyte: b'C', c_uint: b'I', c_ushort: b'S', c_ulong: b'L', c_ulonglong: b'Q',
-        c_float: b'f', c_double: b'd', c_bool: b'B', c_char_p: b'*', c_void_p: b'@',
-        py_object: PyObjectEncoding
-    }
-    return typecodes.get(vartype, b'?')
+_type_to_ctype_map = {
+    int: c_int,
+    float: c_float,
+    bool: c_bool,
+    # This is not really correct - Python strings are Unicode - but we need to keep this for compatibility.
+    str: c_char_p,
+    bytes: c_char_p,
+}
+
+def type_to_ctype(tp):
+    """Convert the given type to a ctypes type.
+    This conversion is applied to types used in objc_method signatures, objc_ivar types, etc.
+    This translates Python built-in types and rubicon.objc classes to their ctypes equivalents.
+    Unregistered types (including things that are already ctypes types) are returned unchanged.
+    """
+    
+    return _type_to_ctype_map.get(tp, tp)
+
+def register_type_to_ctype(tp, ctype):
+    """Register a conversion from a type to a ctypes type."""
+    
+    _type_to_ctype_map[tp] = ctype
+
+def unregister_type_to_ctype(tp):
+    """Unregister a conversion from a type to a ctypes type."""
+    
+    del _type_to_ctype_map[tp]
+
+def get_type_to_ctype_map():
+    """Get a copy of all currently registered type-to-ctype conversions as a mapping."""
+    
+    return dict(_type_to_ctype_map)
+
+_encoding_to_ctype_map = {}
+_ctype_to_encoding_map = {}
+
+def encoding_to_ctype(encoding):
```

Comment:
As a purely stylistic, bikeshed thing, I prefer the `ctype_for_encoding(encoding)` naming scheme, rather than `encoding_to_ctype(encoding)`. To me, the former is easier to read as "this produces a ctype".